### PR TITLE
Export data from BigQuery samples

### PR DESCRIPTION
**Pull request by ace-n**

### CODE REVIEW

File: bigquery/tables.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+'use strict';
+
+// [START auth]
+// By default, gcloud will authenticate using the service account file specified
+// by the GOOGLE_APPLICATION_CREDENTIALS environment variable and use the
+// project specified by the GCLOUD_PROJECT environment variable. See
+// https://googlecloudplatform.github.io/gcloud-node/#/docs/guides/authentication
+var BigQuery = require('@google-cloud/bigquery');
+var Storage = require('@google-cloud/storage');
+var projectId = process.env.GCLOUD_PROJECT;
```
Comment (jmdobry) [inline bigquery/tables.js]: Delete this line. gcloud-node picks up that env var automatically.

File: bigquery/tables.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+// Instantiate the BigQuery and Storage clients
+var bigquery = BigQuery();
+var storage = Storage({
```
Comment (jmdobry) [inline bigquery/tables.js]: Change to `var storage = Storage();`

File: bigquery/tables.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+ */
+function exportTableToGCS (options, callback) {
+  var gcsFileObj = storage.bucket(options.bucket).file(options.file);
+
+  // Export table
+  // See https://googlecloudplatform.github.io/gcloud-node/#/docs/google-cloud/latest/bigquery/table?method=export
+  var table = bigquery.dataset(options.dataset).table(options.table);
+  table.export(
+    gcsFileObj,
+    { format: options.format, gzip: options.gzip },
```
Comment (jmdobry) [inline bigquery/tables.js]: Can we extract this out onto its own line? That way the user can easily see the configuration of the `export` method. e.g.

```js
var table = bigquery.dataset(options.dataset).table(options.table);
var config = {
  format: options.format,
  gzip: options.gzip
};

// Export table
// See https://googlecloudplatform.github.io/gcloud-node/#/docs/google-cloud/latest/bigquery/table?method=export
table.export(gcsFileObj, config, function (err, job) {
```

File: bigquery/tables.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  .command('export <bucket> <file> <dataset> <table> [--format] [--gzip]', 'Export a table from BigQuery to Google Cloud Storage.', {}, function (options) {
+    program.exportTableToGCS(options, console.log);
+  })
+  .command('poll <jobId>', 'Check the status of a BigQuery table export job.', {}, function (options) {
+    program.pollExportJob(options.jobId, console.log);
+  })
+  .example('node $0 export sample-bigquery-export data.json github_samples natality JSON', 'Export github_samples:natality to gcs://sample-bigquery-export/data.json as JSON')
+  .example('node $0 poll job_12345ABCDE', 'Check the status of BigQuery job 12345ABCDE')
+  .options({
+    format: {
```
Comment (jmdobry) [inline bigquery/tables.js]: These aren't global options, so you should move them into the `export` command. Here's an example of how: https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/master/logging/sinks.js#L168

File: bigquery/tables.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  exportTableToGCS: exportTableToGCS,
+  pollExportJob: pollExportJob,
+  main: function (args) {
+    // Run the command-line program
+    cli.help().strict().parse(args).argv;
+  }
+};
+
+cli
+  .command('export <bucket> <file> <dataset> <table> [--format] [--gzip]', 'Export a table from BigQuery to Google Cloud Storage.', {}, function (options) {
```
Comment (jmdobry) [inline bigquery/tables.js]: I'm not sure you want `[--format] [--gzip]` here. I think `yargs` might interpret them as optional positional arguments or as boolean options. `node tables export --help` will generate help text that describes the command's options.

File: bigquery/tables.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    program.exportTableToGCS(options, console.log);
+  })
+  .command('poll <jobId>', 'Check the status of a BigQuery table export job.', {}, function (options) {
+    program.pollExportJob(options.jobId, console.log);
+  })
+  .example('node $0 export sample-bigquery-export data.json github_samples natality JSON', 'Export github_samples:natality to gcs://sample-bigquery-export/data.json as JSON')
```
Comment (jmdobry) [inline bigquery/tables.js]: Maybe show an example that uses the format and gzip options.

File: bigquery/tables.js [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+  })
+  .wrap(100)
+  .recommendCommands()
+  .epilogue('For more information, see https://cloud.google.com/bigquery/exporting-data-from-bigquery');
+
+if (module === require.main) {
+  program.main(process.argv.slice(2), console.log);
+}
+
+module.exports = program;
```
Comment (jmdobry) [inline bigquery/tables.js]: Delete this line, the program has already been exported above.

File: bigquery/tables.js [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+      type: 'boolean',
+      description: 'Whether to compress the exported table using gzip. Defaults to false.'
+    }
+  })
+  .wrap(100)
+  .recommendCommands()
+  .epilogue('For more information, see https://cloud.google.com/bigquery/exporting-data-from-bigquery');
+
+if (module === require.main) {
+  program.main(process.argv.slice(2), console.log);
```
Comment (jmdobry) [inline bigquery/tables.js]: Remove the `console.log` from this line, it's unused in `main`.

File: bigquery/test/tables.test.js [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+
+  var tableMock = {
+    export: sinon.stub().callsArgWith(2, null, jobMock)
+  };
+
+  var datasetMock = {
+    table: sinon.stub().returns(tableMock)
+  };
+
+  var bigqueryMock = {
```
Comment (jmdobry) [inline bigquery/test/tables.test.js]: Collapse the extra lines between all the `var` declarations.

File: bigquery/test/tables.test.js [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+      bigquery: bigqueryMock,
+      Storage: StorageMock,
+      storage: storageMock,
+      metadata: metadataMock,
+      job: jobMock,
+      table: tableMock,
+      bucket: bucketMock,
+      dataset: datasetMock
+    },
+    jobId: jobId
```
Comment (jmdobry) [inline bigquery/test/tables.test.js]: Do you think it's better to put constant variables like this on the result returned by `getSample()`, so they have to be accessed with `sample.jobId`? I've typically just been making my test constants global variables at the top of the file.

File: bigquery/test/tables.test.js [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+      program.exportTableToGCS = sinon.stub();
+
+      program.main(['export', 'bucket', 'file', 'dataset', 'table', 'JSON']);
+      assert(program.exportTableToGCS.calledOnce);
+    });
+
+    it('should call pollExportJob', function () {
+      var program = getSample().program;
+      program.pollExportJob = sinon.stub();
+      var jobId = 'ABCDE';
```
Comment (jmdobry) [inline bigquery/test/tables.test.js]: Re-use the `jobId` var you made near the top of the file.

File: bigquery/test/tables.test.js [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+      assert.equal(callback.firstCall.args[0], error, 'callback received error');
+    });
+  });
+
+  describe('main', function () {
+    it('should call exportTableToGCS', function () {
+      var program = getSample().program;
+      program.exportTableToGCS = sinon.stub();
+
+      program.main(['export', 'bucket', 'file', 'dataset', 'table', 'JSON']);
```
Comment (jmdobry) [inline bigquery/test/tables.test.js]: If any of these magic strings appear elsewhere in this file, create a single var at the top of the file that you can re-use throughout the file.

File: bigquery/tables.js (hunk 6 above)
Comment (ace-n) [inline bigquery/tables.js]: Should I add a test as well?

File: bigquery/test/tables.test.js (hunk 10 above)
Comment (ace-n) [inline bigquery/test/tables.test.js]: One benefit of doing it this way is that everything related to dummy values is in a single package - but I'll change over to your way for consistency with the other samples.

File: bigquery/test/tables.test.js (hunk 9 above)
Comment (ace-n) [inline bigquery/test/tables.test.js]: Fixed

File: bigquery/tables.js (hunk 8 above)
Comment (ace-n) [inline bigquery/tables.js]: Fixed

File: bigquery/tables.js (hunk 7 above)
Comment (ace-n) [inline bigquery/tables.js]: Fixed

File: bigquery/tables.js (hunk 4 above)
Comment (ace-n) [inline bigquery/tables.js]: Fixed

File: bigquery/tables.js (hunk 3 above)
Comment (ace-n) [inline bigquery/tables.js]: Fixed

File: bigquery/tables.js (hunk 2 above)
Comment (ace-n) [inline bigquery/tables.js]: Fixed

File: bigquery/tables.js (hunk 1 above)
Comment (ace-n) [inline bigquery/tables.js]: Fixed

File: bigquery/tables.js (hunk 6 above)
Comment (jmdobry) [inline bigquery/tables.js]: Yeah, somehow we want to make sure they work.

File: bigquery/system-test/tables.test.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+'use strict';
+
+var uuid = require('node-uuid');
+var generateUuid = function () {
```
Comment (ace-n) [inline bigquery/system-test/tables.test.js]: Hopefully doing this as a function (for brevity) is fine.

File: bigquery/system-test/tables.test.js [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+var options = {
+  bucket: generateUuid(),
+  file: 'data.json',
+  dataset: generateUuid(),
+  table: generateUuid()
+};
+var BigQuery = require('@google-cloud/bigquery');
+var bigquery = BigQuery();
+var Storage = require('@google-cloud/storage');
+var storage = Storage();
```
Comment (ace-n) [inline bigquery/system-test/tables.test.js]: Should these be compressed, i.e `var bigquery = require('@google-cloud/bigquery')();`?

Comment (jmdobry) [inline bigquery/system-test/tables.test.js]: It's fine either way.

File: bigquery/system-test/tables.test.js [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+      });
+    });
+  });
+  after(function (done) {
+    // Delete table export
+    file.delete(function () {
+      // Delete testing dataset/table
+      bigquery.dataset(options.dataset).delete({ force: true }, done);
+    });
+  });
```
Comment (jmdobry) [inline bigquery/system-test/tables.test.js]: This test isn't deleted the bucket that it creates. The simplest way to clean up a bucket and all its files can be found here: https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/master/storage/system-test/files.test.js#L41

File: bigquery/tables.js [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+ * @param {string} options.format Format to export as - either 'CSV', 'JSON', or 'AVRO'.
+ * @param {boolean} [options.gzip] Optional. Whether or not data should be compressed using GZIP.
+ * @param {function} callback Callback function to receive query results.
+ */
+function exportTableToGCS (options, callback) {
+  var gcsFileObj = storage.bucket(options.bucket).file(options.file);
+  // Export table
+  // See https://googlecloudplatform.github.io/gcloud-node/#/docs/google-cloud/latest/bigquery/table?method=export
+  var table = bigquery.dataset(options.dataset).table(options.table);
+  table.export(
```
Comment (jmdobry) [inline bigquery/tables.js]: This is a little hard to read, can we do:

```js
var config = {
  format: options.format,
  gzip: options.gzip
};
table.export(gcsFileObj, config, function (err, job) {
```

File: bigquery/tables.js [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+
+cli
+  .command('export <bucket> <file> <dataset> <table>', 'Export a table from BigQuery to Google Cloud Storage.', {
+    format: {
+      alias: 'f',
+      global: true,
+      requiresArg: true,
+      type: 'string',
+      choices: ['JSON', 'CSV', 'AVRO'],
+      default: 'JSON'
```
Comment (jmdobry) [inline bigquery/tables.js]: I don't think you should have a default here, `@google-cloud/bigquery` has its own default that it chooses (which happens to be "csv").

File: bigquery/tables.js [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+    gzip: {
+      global: true,
+      type: 'boolean',
+      description: 'Whether to compress the exported table using gzip. Defaults to false.'
+    }
+  }, function (options) {
+    program.exportTableToGCS(options, console.log);
+  })
+  .example(
+    'node $0 export sample-bigquery-export data.json github_samples natality',
```
Comment (jmdobry) [inline bigquery/tables.js]: Can you change the names in these samples to be recognizable as what they are? For example, `github_samples` could be a bucket, dataset, table, or I don't know what. I'd have to look at the positional signature of the command to figure it out.

Better for these examples would be to have `node $0 export my-bucket my-file.json my_dataset my_table`, then it becomes obvious to the user what is what.

File: bigquery/tables.js [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+    'Export github_samples:natality to gcs://sample-bigquery-export/data.json as raw JSON'
+  )
+  .example(
+    'node $0 export sample-bigquery-export data.csv github_samples natality -f CSV --gzip',
+    'Export github_samples:natality to gcs://sample-bigquery-export/data.csv as gzipped CSV'
+  )
+  .wrap(100)
+  .recommendCommands()
+  .epilogue('For more information, see https://cloud.google.com/bigquery/exporting-data-from-bigquery');
```
Comment (jmdobry) [inline bigquery/tables.js]: Might want to change the link to just `https://cloud.google.com/bigquery/docs` as this file will contain other examples unrelated to exporting data from BigQuery.

File: bigquery/tables.js (hunk 16 above)
Comment (jmdobry) [inline bigquery/tables.js]: Or better yet

```js
var gcsFileObj = storage.bucket(options.bucket).file(options.file);
var table = bigquery.dataset(options.dataset).table(options.table);

var config = {
  format: options.format,
  gzip: options.gzip
};

See https://googlecloudplatform.github.io/gcloud-node/#/docs/google-cloud/latest/bigquery/table?method=export
table.export(gcsFileObj, config, function (err, job) {
```

File: bigquery/system-test/tables.test.js (hunk 15 above)
Comment (ace-n) [inline bigquery/system-test/tables.test.js]: Fixed.

File: bigquery/tables.js (hunk 16 above)
Comment (ace-n) [inline bigquery/tables.js]: Fixed.

File: bigquery/tables.js (hunk 18 above)
Comment (ace-n) [inline bigquery/tables.js]: Fixed.

File: bigquery/tables.js (hunk 17 above)
Comment (ace-n) [inline bigquery/tables.js]: Hmm...when I tested it, the default seemed to be JSON. (Though I agree with the "`@google-cloud/bigquery` has its own default" part.)

File: bigquery/tables.js (hunk 19 above)
Comment (ace-n) [inline bigquery/tables.js]: Fixed.

File: bigquery/system-test/tables.test.js [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+            assert.ifError(err, 'table creation succeeded');
+            done();
+          }
+        );
+      });
+    });
+  });
+  after(function (done) {
+    // Delete table export
+    file.delete(function (err) {
```
Comment (jmdobry) [inline bigquery/system-test/tables.test.js]: You don't need this line, this file will be deleted by the `deleteFiles` 10 lines down.

File: bigquery/system-test/tables.test.js [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+      if (err) {
+        return done(err);
+      }
+      // Delete testing dataset/table
+      bigquery.dataset(options.dataset).delete({ force: true }, function (err) {
+        if (err) {
```
Comment (jmdobry) [inline bigquery/system-test/tables.test.js]: I would ignore this error, you still want to attempt to delete the bucket even if deleting the dataset failed for some reason.

If you want you could keep track of the error, then pass it to `done` after deleting the bucket.

File: bigquery/system-test/tables.test.js (hunk 20 above)
Comment (ace-n) [inline bigquery/system-test/tables.test.js]: Whoops, good catch. Will fix.

File: bigquery/system-test/tables.test.js (hunk 21 above)
Comment (ace-n) [inline bigquery/system-test/tables.test.js]: Will fix.

**Comment by jmdobry** [comment]
LGTM, will merge when green.

**Pull request merged**